### ISSUE TITLE
training: hard-case-focused predictive retraining with crossing-conflict augmentation

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#3214]
## Archetype Metadata

```yaml
archetype: training-campaign
evidence_tier: idea
linked_policy:
  - docs/maintainer_values.md
  - configs/benchmarks/predictive_hard_seeds_v1.yaml
```

## Goal / Problem

The third bet in the hard-case breakthrough portfolio (#C) is **model-side**: the predictive model may
not anticipate crossing-conflict interactions well, so hard cases fail regardless of selection (#3204)
or maneuver authority (sibling issue). This issue tests whether **hard-case-focused retraining / data
augmentation** — oversampling or augmenting crossing-conflict phases in the trajectory-model training
set — raises hard-seed benchmark success.

Runs in parallel with the other two bets; independent so the portfolio can be a true tournament.

## Hypothesis / Claim contract

- **Target hypothesis:** retraining the predictive model with crossing-conflict-weighted data raises
  hard-seed success over the current `predictive_proxy_selected_v1` lineage.
- **Comparator / baseline:** current model lineage on `predictive_hard_seeds_v1`, same algo YAML + seeds.
- **Minimum valid evidence:** ≥ 1 retrained checkpoint vs baseline on the hard-seed benchmark, with the
  navigation gate kept separate from the trajectory gate (`val_ade`/`val_fde`), per existing policy.
- **Decision / stop rule:** adopt if hard success improves with non-regressing overall success; if ADE
  improves but hard success does not, that corroborates the #3204 finding that ADE ≠ benchmark success
  and is recorded as such. Negative result if neither moves.
- **Provenance plan:** training config + data-weighting spec under `configs/`; checkpoint promoted via
  `model/registry.yaml` with W&B run-file metadata (no reliance on `output/tmp`); context note + summary.

## Scope / Out of scope

- In: crossing-conflict-weighted sampling/augmentation in the predictive-model training pipeline; one
  retrained checkpoint; hard-seed evaluation.
- Out: new model architectures (tracked separately, e.g. #2844); paper-grade claim before portfolio
  synthesis (#C).

## Acceptance Criteria

- [ ] Hard-case-weighted training config + data-weighting spec are explicit and reproducible.
- [ ] ≥ 1 retrained checkpoint evaluated vs baseline on `predictive_hard_seeds_v1`; nav gate and ADE
      gate reported separately.
- [ ] Decision per stop rule with uncertainty; checkpoint promoted durably (registry/W&B), not local-only.
- [ ] Result classified on the evidence ladder; no fallback/degraded promotion.

## Validation / Testing

```bash
uv run python scripts/training/train_predictive_model.py --config configs/research/predictive_hardcase_retrain_v1.yaml   # name TBD by implementer
uv run python scripts/validation/evaluate_predictive_planner.py --algo-config <retrained-algo>.yaml --benchmark configs/benchmarks/predictive_hard_seeds_v1.yaml
```

## Workflow

Branch/worktree + online PR only. Suggested: `../robot_sf_ll7.worktrees/issue-NNN-hardcase-retrain`.

## Estimate

Effort: 16–28h + training compute (SLURM). Risk: medium. Resource: SLURM for the training run.

## Blocked by

Needs the hard-case training data path; coordinate with existing trace/data lanes. Portfolio sibling of
#3204; parent = hard-case portfolio epic.

## 🤖 Agent execution spec (codex gpt5.5 medium)

_Appended 2026-06-20. Original body unchanged. Agent runtime: local only. **Config + data-weighting + augmentation code are local**; the retraining run needs SLURM/GPU._

`archetype: training-campaign` · `evidence_tier: blocked` → `nominal` after the retrained checkpoint exists.

### Verified entry points
- Predictive training pipeline: `scripts/training/train_predictive_planner.py`, `scripts/training/run_predictive_training_pipeline.py`.
- Hard-case data: `scripts/training/collect_predictive_hardcase_data.py`, `scripts/training/build_predictive_mixed_dataset.py`.
- Eval: `scripts/validation/evaluate_predictive_planner.py` on `configs/benchmarks/predictive_hard_seeds_v1.yaml`.
- Promotion: `model/registry.yaml` (no `output/tmp` reliance).

### Agent-executable slice (local)
1. Write the crossing-conflict-weighted training config + explicit data-weighting/augmentation spec under `configs/` (oversample/augment crossing-conflict phases in the trajectory-model training set).
2. Implement the augmentation in `build_predictive_mixed_dataset.py` (or a sibling) with a unit test on a small fixture.
3. Keep the navigation gate separate from the trajectory gate (`val_ade`/`val_fde`) per existing policy.

### Acceptance criteria (mirrors body)
- [ ] Hard-case-weighted training config + data-weighting spec explicit + reproducible.
- [ ] ≥1 retrained checkpoint vs baseline on `predictive_hard_seeds_v1` (nav gate ≠ ADE gate).
- [ ] ADE-up-but-hard-success-flat corroborates #3204 and is recorded as such; negative result if neither moves.

### Validation commands
```bash
uv run python scripts/training/build_predictive_mixed_dataset.py --help
uv run python -m pytest tests/ -k "predictive and (dataset or augment or hard)" -q
```

### Blocked-until
- The retrain run is SLURM/GPU (`resource:slurm`). The agent prepares config + augmentation + dataset build + eval harness; SLURM produces the checkpoint.

**Comment by ll7** [comment]
Portfolio sibling under epic #3215 (model bet).

**Comment by ll7** [comment]
Scoped launch/config update merged in PR #3255:

- added the #3214 crossing-conflict hard-case weighting spec;
- added pipeline support and tests for resolving `mixing.weighting_spec`;
- opened follow-up issue #3254 for the actual weighted retraining, checkpoint provenance, and hard-seed evaluation evidence.

This does **not** publish a retrained checkpoint or model-improvement result for #3214. Since #3254 now owns the evidence-producing follow-up, I am releasing the temporary agent claim for the #3214 tooling/config slice.

**Comment by ll7** [comment]
**Maintainer priority note** (via `/issue-audit`, 2026-06-30): **sequenced after #3216** for the paper-critical SLURM lane. #3216 (paper-grade headline planner comparison) takes the next slot to lock down headline numbers / rank stability first; this retraining job follows. Note: completing this may make a #3216 rerun worthwhile. Both remain `priority: high`; this records execution order only.

**Comment by ll7** [comment]
PR #4467 merged: added packet/pipeline hard-seed summary consistency validation (`evaluation_config.summary_path` must match `output.provenance.hard_seed_evaluation_summary`, fails closed on divergence). Static readiness-contract progression only — #3214 remains `retraining_blocked` (still sequenced after #3216; needs the control-law-side change + checkpoint/evaluation provenance plan). No compute submitted.

**Comment by ll7** [comment]
BLOCKED: The issue is currently blocked per the latest maintainer decision because it is sequenced after #3216, requires the control-law-side change, and needs a checkpoint/evaluation provenance plan.

**Comment by ll7** [comment]
BLOCKED: This issue is blocked per maintainer decision as it is sequenced after #3216, requires a control-law-side change, and needs a checkpoint/evaluation provenance plan.

**Comment by ll7** [comment]
BLOCKED: Cannot produce a new implementation slice. The issue is blocked per maintainer decision — sequenced after #3216 (paper-grade headline planner comparison), requires the control-law-side change, and needs a checkpoint/evaluation provenance plan. Two prior PRs (#3255 config/tooling, #4467 hard-seed validation) already cover the non-evidence tooling slices. No further code slice is possible until the blockers resolve.

**Comment by ll7** [comment]
BLOCKED (cheap-lane implementation worker, 2026-07-04): no new non-duplicative implementation slice is possible for #3214. Verified against current repo state, not only the prior thread:

- **Sequencing predecessor is closed.** The named execution-order predecessor #3216 (paper-grade headline planner comparison) is now `CLOSED`, so that ordering constraint is satisfiable — but it is not the binding blocker.
- **Empirical block (the binding reason).** `configs/training/predictive/predictive_retraining_readiness_issue_3214.yaml` records the #3254 crossing-conflict weighted-retrain result as `status: verified_negative` with `observed_success_rate: 0.087` against `gate_min_success_rate: 0.30` (PR #3515). The retrain improved trajectory-prediction accuracy but did not clear the closed-loop policy gate, corroborating the #3204 ADE≠benchmark-success finding. The packet sets `launch_decision.state: blocked_until_control_law_change` and `compute_submit_allowed: false`.
- **Unmet required inputs** (packet `required_before_rerun`): `control_law_change_config`, `updated_public_config_or_pr`, `checkpoint_provenance_plan`, `hard_seed_evaluation_plan`. A repo search finds `control_law_change_config` only as the *prerequisite name* inside the packet, its validator `robot_sf/training/predictive_retraining_readiness.py`, and tests — the prerequisite is named, not satisfied. The `targeted_control_law_hardening` reference lives in a different issue's exploratory config (#4205) and is not wired into this packet as a satisfied prerequisite.
- **Prior tooling slices already merged** and not to be duplicated: PR #3255 (crossing-conflict weighting spec + `mixing.weighting_spec` resolution + `repeat_hardcase_rows` oversampling in `build_predictive_mixed_dataset.py` + unit tests) and PR #4467 (`evaluation_config.summary_path` ↔ `output.provenance.hard_seed_evaluation_summary` hard-seed summary consistency validation). Adding further augmentation/repeat variants would be same-packet retraining tooling that the readiness packet and the prior BLOCKED decisions explicitly hold until a control-law-side change exists.

This worker produced no code and submitted no compute (no Slurm, no training, no benchmark campaign). Unblock path: land a public control-law-side change plus checkpoint/evaluation provenance plan, then re-open the retrain slice.

**Comment by ll7** [comment]
State update (gate): PR #4568 merged — added durable closure-audit evidence note (`docs/context/evidence/issue_3214_closure_audit_2026-07-05.md`) mapping acceptance criteria to merged PRs #3255/#3515/#3869/#4467 + follow-up #3254. Issue remains OPEN (`blocked_not_closable`): still needs a public control-law-side change plus checkpoint-provenance and hard-seed evaluation plans before closure. Docs-only; no runtime/benchmark/model change.
